11 - Custom Video Player/scripts.js

```diff
@@ -0,0 +1,103 @@
+class CustomVideoPlayer {
+    constructor(selector) {
+        this.player = document.querySelector(selector);
+        this.video = this.player.querySelector('.viewer');
+        this.progress = this.player.querySelector('.progress');
+        this.progressBar = this.player.querySelector('.progress__filled');
+        this.toggle = this.player.querySelector('.toggle');
+        this.skipButtons = this.player.querySelectorAll('[data-skip]');
+        this.ranges = this.player.querySelectorAll('.player__slider');
+        this.fullScreenToggle = this.player.querySelector('.fullScreenToggle');
+        this.mouseDown = false;
+    };
+
+    playPause = () => {
+        if (this.video.paused) this.video.play();
+        else this.video.pause();
+      };
+
+    toggleIcon = (evt) => {
+        this.toggle.textContent = evt.target.paused ? '►' : '❚ ❚';
+    };
+    
+    skipSeconds = (evt) => {
+        this.video.currentTime += parseFloat(evt.target.dataset.skip);
+    };
+
+    handleRanges = (evt) => {
+        const { name, value } = evt.target;
+        this.video[name] = value;
+    };
+
+    handleTime = (evt) => {
+        const progressPercent = (this.video.currentTime / this.video.duration) * 100;
+        this.progressBar.style.flexBasis = `${progressPercent}%`;
+    };
+
+    dragProgress = (evt) => {
+        const scrubTime = (evt.offsetX / this.progress.offsetWidth) * this.video.duration;
+        this.video.currentTime = scrubTime;
+      };
+
+    toggleMouseDown = () => {
+        this.mouseDown = !this.mouseDown;
+    };
+
+    handleFullScreen = () => {
+        document.fullscreenElement = 
+            document.fullscreenElement || 
+            document.mozFullscreenElement || 
+            document.msFullscreenElement || 
+            document.webkitFullscreenDocument;
+
+        document.exitFullscreen = 
+            document.exitFullscreen || 
+            document.mozExitFullscreen || 
+            document.msExitFullscreen || 
+            document.webkitExitFullscreen;
+
+        this.video.requestFullscreen = 
+            this.video.requestFullscreen || 
+            this.video.mozRequestFullscreen || 
+            this.video.msRequestFullscreen || 
+            this.video.webkitRequestFullscreen;
+
+        
+        if (!document.fullscreenElement) {
+            this.video.requestFullscreen()
+                .catch(err => {
+                    alert(`Error attempting to enable full-screen mode: ${err.message} (${err.name})`);
+                });
+        } else {
+          if (document.exitFullscreen) {
+            document.exitFullscreen();
+          }
+        }
+        
+    };
+};
+
+//// class init
+const newVideoPlayer = new CustomVideoPlayer('.player');
+
+//// listeners
+// play-pause handlers
+newVideoPlayer.video.addEventListener('click', newVideoPlayer.playPause);
+newVideoPlayer.toggle.addEventListener('click', newVideoPlayer.playPause);
```

Comment:
here you are accessing the „internals“ of the class from the outside. i would consider this as a breach of the encapsulation principle. i tend to think about it this way: there is a (well-documented) public interface of a class. this has to be stable and developers can use it. however there is also a private part (i would consider these „selector cache elements“ as such. these private parts can be changed by the class owner at any time. hence, you shall not rely on them from the outside.

what do you think?

Reply:
Just for me to better understand your point... do you mean that basically all `eventListeners` should be added as `newVideoPlayer.player.addEventListener()`? I tried adding conditions to the `playPause` method (such as `if (evt.target === this.video || evt.target === this.toggle)`. Is this the best way to do it? Just to know before I change all `eventListeners`.

I think your point is totally right but sometimes is really challenging because from the `player` we need to get to the event trigger or target within it.

Reply:
it is really hard to explain. I basically try to look at a class as a black box which only exposes the minimum of methods/variables required.

think about it like a oven/stove. a modern oven is complicated, but to the outside it only has a few buttons and you as the user do not need to worry about the internals. here with `newVideoPlayer.video.addEventListener` you are basically doing `oven.internalWiring.doSomething`. This is not nice and the oven manufacturer may decided to change its internal wiring without notifying you.

so, I guess what I am trying to say is, maybe move the registration of event listeners into the class :)